base/deprecated.jl

```diff
@@ -56,41 +56,33 @@ macro deprecate(old,new)
     end
 end
 
+const have_dep_warned = Set()
+
 function depwarn(msg, funcsym)
     opts = JLOptions()
-    if opts.depwarn > 0
-        ln = Int(unsafe_load(cglobal(:jl_lineno, Cint)))
+    if opts.depwarn == 1  # raise a warning
         fn = String(unsafe_load(cglobal(:jl_filename, Ptr{Cchar})))
+        ln = Int(unsafe_load(cglobal(:jl_lineno, Cint)))
+        key = hash(string(fn, ln, msg))
+        (key in have_dep_warned) && return
```

Comment:
@yuyichao I was wondering about than myself. In practice the `have_dep_warned` filter seems less granular than the caller filter below.

My ad hoc testing of comparing the before and after output has shown that the same amount of deprecation warnings are emitted. 

So far everything is just proof-of-concept and I'll need to figure out some good tests. Could you provide me with a test that would cover your concern?